In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as Plt 
import seaborn as sns 
print("Libraries imported successfully!")

Libraries imported successfully!


In [5]:
import os
file_path = "../dataset/ Titanic-Dataset.csv"
print("File exists:",os.path.exists(file_path))
print("Files in dataset folder:",os.listdir("../dataset"))

File exists: False
Files in dataset folder: ['Titanic-Dataset.csv']


In [7]:
# Step 2: Data Quality Report

import pandas as pd

# Load dataset
file_path = "../dataset/Titanic-Dataset.csv"
df = pd.read_csv(file_path)

print("Dataset loaded successfully!")
print("Dataset shape:", df.shape)

# Dataset information
print("\n--- Dataset Information ---")
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

# Column names
print("\n--- Column Names ---")
print(df.columns.tolist())

# Data types
print("\n--- Data Types ---")
display(df.dtypes)

# Missing values
print("\n--- Missing Values ---")
missing_values = df.isnull().sum()
display(missing_values)

# Duplicate rows
print("\n--- Duplicate Rows ---")
duplicate_count = df.duplicated().sum()
print("Number of duplicate rows:", duplicate_count)

# Numerical summary
print("\n--- Numerical Summary ---")
display(df.describe())

Dataset loaded successfully!
Dataset shape: (891, 12)

--- Dataset Information ---
Number of rows: 891
Number of columns: 12

--- Column Names ---
['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']

--- Data Types ---


PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object


--- Missing Values ---


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


--- Duplicate Rows ---
Number of duplicate rows: 0

--- Numerical Summary ---


,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [8]:
# Step 3: Missing Values Analysis

print("--- Missing Values Analysis ---")

missing_values = df.isnull().sum()
missing_percentage = (missing_values / len(df)) * 100

missing_report = pd.DataFrame({
    "Missing Values": missing_values,
    "Missing Percentage": missing_percentage.round(2)
})

display(missing_report)

--- Missing Values Analysis ---


,Missing Values,Missing Percentage
PassengerId,0,0.00
Survived,0,0.00
Pclass,0,0.00
Name,0,0.00
Sex,0,0.00
Age,177,19.87
SibSp,0,0.00
Parch,0,0.00
Ticket,0,0.00
Fare,0,0.00


In [9]:
# Step 4: Duplicate Row Analysis

print("--- Duplicate Row Analysis ---")

duplicate_count = df.duplicated().sum()

print("Number of duplicate rows:", duplicate_count)

--- Duplicate Row Analysis ---
Number of duplicate rows: 0


### Missing Data Handling

The missing-value analysis identified missing data in the Age, Cabin, and Embarked columns. The Age column contains 177 missing values (19.87%), so median imputation is used because Age is a numerical variable and the median is less sensitive to extreme values. The Cabin column contains 687 missing values (77.10%), which is a very high proportion; therefore, the column is removed because retaining it would provide limited reliable information. The Embarked column contains only 2 missing values (0.22%), so mode imputation is applied because it is a categorical variable with very few missing observations.

In [10]:
# Step 5: Missing Data Handling

# Age: Fill missing values with median
age_median = df['Age'].median()
df['Age'] = df['Age'].fillna(age_median)

# Embarked: Fill missing values with mode
embarked_mode = df['Embarked'].mode()[0]
df['Embarked'] = df['Embarked'].fillna(embarked_mode)

# Cabin: Drop because 77.10% of values are missing
df = df.drop(columns=['Cabin'])

print("Missing data handling completed successfully!")
print("Age median used:", age_median)
print("Embarked mode used:", embarked_mode)
print("Cabin column removed due to high missing percentage.")

print("\nMissing values after handling:")
display(df.isnull().sum())

Missing data handling completed successfully!
Age median used: 28.0
Embarked mode used: S
Cabin column removed due to high missing percentage.

Missing values after handling:


PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

In [11]:
# Step 6: Duplicate Row Removal

duplicates_before = df.duplicated().sum()

print("Duplicate rows before cleaning:", duplicates_before)

df = df.drop_duplicates()

duplicates_after = df.duplicated().sum()

print("Duplicate rows after cleaning:", duplicates_after)
print("Duplicate removal completed successfully!")
print("Dataset shape after duplicate removal:", df.shape)

Duplicate rows before cleaning: 0
Duplicate rows after cleaning: 0
Duplicate removal completed successfully!
Dataset shape after duplicate removal: (891, 11)


### Data Standardisation

The categorical values were checked for inconsistent formatting. The Sex column was standardised to lowercase and then converted to a consistent format. The Embarked column was also checked for unique categorical values. Identifier fields such as PassengerId and Ticket were retained as identifiers rather than treated as numerical measurements.

In [12]:
# Step 7: Data Standardisation

# Standardise Sex values
df['Sex'] = df['Sex'].astype(str).str.strip().str.lower()

# Standardise Embarked values
df['Embarked'] = df['Embarked'].astype(str).str.strip().str.upper()

# Convert PassengerId to string because it is an identifier
df['PassengerId'] = df['PassengerId'].astype(str)

print("Data standardisation completed successfully!")

print("\nUnique Sex values:")
print(df['Sex'].unique())

print("\nUnique Embarked values:")
print(df['Embarked'].unique())

print("\nUpdated data types:")
display(df.dtypes)

Data standardisation completed successfully!

Unique Sex values:
['male' 'female']

Unique Embarked values:
['S' 'C' 'Q']

Updated data types:


PassengerId     object
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Embarked        object
dtype: object

In [13]:
# Check Embarked values for possible anomalies

print("Embarked value counts:")
display(df['Embarked'].value_counts(dropna=False))

Embarked value counts:


Embarked
S    646
C    168
Q     77
Name: count, dtype: int64

### Outlier Detection Using IQR

The Interquartile Range (IQR) method was used to identify potential outliers in the numerical variables. The IQR is calculated as the difference between the third quartile (Q3) and the first quartile (Q1). Values below Q1 − 1.5 × IQR or above Q3 + 1.5 × IQR are identified as potential outliers. The analysis is performed on Age and Fare, while identifier and categorical variables are excluded.

In [14]:
# Step 8: Outlier Detection Using IQR

numeric_columns = ['Age', 'Fare']

outlier_summary = []

for column in numeric_columns:
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = df[
        (df[column] < lower_bound) |
        (df[column] > upper_bound)
    ]

    outlier_summary.append({
        'Column': column,
        'Q1': Q1,
        'Q3': Q3,
        'IQR': IQR,
        'Lower Bound': lower_bound,
        'Upper Bound': upper_bound,
        'Outlier Count': len(outliers)
    })

outlier_report = pd.DataFrame(outlier_summary)

print("Outlier detection completed successfully!")
display(outlier_report)

Outlier detection completed successfully!


,Column,Q1,Q3,IQR,Lower Bound,Upper Bound,Outlier Count
0,Age,22.0000,35.0,13.0000,2.500,54.5000,66
1,Fare,7.9104,31.0,23.0896,-26.724,65.6344,116


### Outlier Treatment Decision

The IQR method identified 66 potential outliers in Age and 116 potential outliers in Fare. These observations were retained because the extreme values may represent genuine passenger characteristics, such as older passengers or passengers who paid higher fares. Removing them could result in loss of meaningful information. Therefore, the identified outliers were documented but not removed from the cleaned dataset.

In [15]:
# Step 9: Final Data Quality Check

print("--- Final Data Quality Check ---")

print("Final dataset shape:", df.shape)

print("\nMissing values after cleaning:")
display(df.isnull().sum())

print("\nDuplicate rows after cleaning:")
print(df.duplicated().sum())

print("\nFinal data types:")
display(df.dtypes)

print("\nFinal numerical summary:")
display(df.describe())

--- Final Data Quality Check ---
Final dataset shape: (891, 11)

Missing values after cleaning:


PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64


Duplicate rows after cleaning:
0

Final data types:


PassengerId     object
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Embarked        object
dtype: object


Final numerical summary:


,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.361582,0.523008,0.381594,32.204208
std,0.486592,0.836071,13.019697,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,22.000000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,35.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [16]:
# Step 10: Save Cleaned Dataset

output_path = "../output/Titanic_Cleaned.csv"

df.to_csv(output_path, index=False)

print("Cleaned dataset saved successfully!")
print("File path:", output_path)
print("Final dataset shape:", df.shape)

Cleaned dataset saved successfully!
File path: ../output/Titanic_Cleaned.csv
Final dataset shape: (891, 11)
